app/group.go

```diff
@@ -86,30 +88,51 @@ func (a *App) UpsertGroupSyncable(groupSyncable *model.GroupSyncable) (*model.Gr
 		return nil, err
 	}
 
-	if gs == nil {
-		gs, err = a.Srv().Store.Group().CreateGroupSyncable(groupSyncable)
+	// reject the syncable creation if the group isn't already associated to the parent team
+	if groupSyncable.Type == model.GroupSyncableTypeChannel {
+		var channel *model.Channel
+		channel, err = a.Srv().Store.Channel().Get(groupSyncable.SyncableId, true)
 		if err != nil {
 			return nil, err
 		}
-	} else {
-		gs, err = a.Srv().Store.Group().UpdateGroupSyncable(groupSyncable)
+
+		var team *model.Team
+		team, err = a.Srv().Store.Team().Get(channel.TeamId)
 		if err != nil {
 			return nil, err
 		}
+		if team.IsGroupConstrained() {
+			var teamGroups []*model.Group
+			teamGroups, err = a.GetAllGroupsByTeam(team.Id)
+			if err != nil {
+				return nil, err
+			}
+			var permittedGroup bool
+			for _, teamGroup := range teamGroups {
+				if teamGroup.Id == groupSyncable.GroupId {
+					permittedGroup = true
+					break
+				}
+			}
+			if !permittedGroup {
+				return nil, model.NewAppError("App.UpsertGroupSyncable", "group_not_associated_to_synced_team", nil, "", http.StatusBadRequest)
+			}
+		} else {
+			_, err = a.UpsertGroupSyncable(model.NewGroupTeam(groupSyncable.GroupId, team.Id, groupSyncable.AutoAdd))
```

Comment:
I don't understand why are we recurring here. since it is not a `GroupSyncableTypeChannel` it will ignore the first part of the `if`, and all the other has already been done, so it would only be iterating.

or maybe I'm missing some pieces, as I don't know this part of code.

Reply:
I think this whole block needs a refactor given it's confusing to read.

Line 92 branches this entire diff, meaning that all of these changes related only to a "group channel" (aka `GroupSyncableTypeChannel`) upsert request.

The requirements can be summarized like this:

* Existing requirement: if a parent team is not group-constrained, then create the group channel *and* the group-team.
* New requirement (that was labelled a bug, but really it's a requirement change): if the parent team is group-constrained don't create anything unless the team is already associated to the group.